### PR TITLE
Enforce SSO control at signup / login

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

This PR goes one step further than https://github.com/dust-tt/dust/pull/4227 by enforcing SSO during login/signup. If SSO is required for a matching workspace, non-sso accounts will be redirected to the SSO login flow when signing up or logging in. This is needed to short circuit user creation and all related update on a workspace pricing.  

## Deploy Plan

- [x] Whitelist logout url for `sso-enforced`.

### CODE REVIEW

**Review by spolu — COMMENTED** [review]
Our invariant is that if error needs to be catched they should be retuned as a Result

Can you refactor accordingly?

**Comment by flvndvd** [comment]
> Our invariant is that if error needs to be catched they should be retuned as a Result
> 
> Can you refactor accordingly?

I know but in this scenario, however, it leads to substantial code duplication, as we must repeatedly check for and handle `SSOEnforcedError` across all sign-up/login methods where it might occur. Let me write some more lines of code then.

**Comment by flvndvd** [comment]
> Our invariant is that if error needs to be catched they should be retuned as a Result
> 
> Can you refactor accordingly?

Does your comment also target the `FrontApiError`?

**Comment by spolu** [comment]
Concerning Front error: if it’s handled specifically and we control that error then yes.

The invariant is: throwing means unrecoverable error, just 500. Anything else should use Result 👍

**Review by fontanierh — APPROVED** [review]
LGTM

File: front/pages/api/login.ts [hunk 1]
```diff
@@ -297,11 +334,27 @@ async function handler(
...
-        user,
-        targetWorkspaceId
-      );
+      const loginFctn = membershipInvite
+        ? async () => handleMembershipInvite(user, membershipInvite)
+        : async () => handleRegularSignupFlow(session, user, targetWorkspaceId);
+
+      const result = await loginFctn();
+      if (result.isErr()) {
+        if (result.error instanceof FrontApiError) {
```
Comment (spolu) [inline front/pages/api/login.ts:343]: If we rethrow, we can probably throw direclty all FrontAPIerror?

**Review by spolu — APPROVED** [review]
Left a small comment about FrontApiError

File: front/pages/api/login.ts [hunk 2]
```diff
@@ -273,52 +302,67 @@ async function handler(
...
-      const { flow, workspace } = await handleRegularSignupFlow(
-        session,
-        user,
-        targetWorkspaceId
-      );
-      if (flow) {
-        res.redirect(`/no-workspace?flow=${flow}`);
-        return;
+      if (userCreated) {
+        await user.destroy();
```
Comment (spolu) [inline front/pages/api/login.ts:350]: Not sure I understand this one. Maybe worth a comment?

**Review by spolu — APPROVED** [review]
Thanks for updating error handling \o/

Hope you're aligned with this setup :+1:

**Comment by flvndvd** [comment]
> Thanks for updating error handling \o/
> 
> Hope you're aligned with this setup 👍

Yep aligned 👌

**Pull request merged**